csrc/python_frontend/fusion_definition.cpp

```diff
@@ -497,10 +513,12 @@ std::pair<KernelArgumentHolder, std::vector<Sharding>> FusionDefinition::
 
   std::vector<Sharding> output_shardings;
   if (user_sched == nullptr) {
-    FusionKernelRuntime* runtime =
-        scheds->auto_gen_schedules->getMostRecentKernelRuntime();
-    output_shardings =
-        getOutputShardings(runtime->fusionSegments()->completeFusion());
+    Fusion* fusion = use_multidevice_executor_
+        ? scheds->preschedFusion()
```

Comment:
FYI, using preschedFusion for MultiDeviceExecutor means that its user has to specify multi-GPU schedules for all output tensors. That's certainly a valid contract -- just to make sure you are aware of that implication. 

FusionExecutorCache follows a different contract. Sharding propagation will try to complete all output schedules if they are missing a DeviceMesh. Therefore, completeFusion was used instead of preschedFusion. 

Both contracts have pros and cons. I'm sure we'll form a better opinion when more clarity goes into Thunder/TP integration.

Reply:
Ok thanks, I did not have this in mind, that's useful info. For now I'm ok with the contract that user needs to define the full schedule for `MultiDeviceExecutor`. We'll see if/when we'll add sharding propagation in `MultiDeviceExecutor`.